adam-hbase/src/main/scala/org/bdgenomics/adam/hbase/HBaseFunctions.scala

```diff
@@ -0,0 +1,627 @@
+                              optConnection: Option[Connection] = None,
+                              optAdmin: Option[Admin] = None,
+                              optHBaseContext: Option[HBaseContext] = None) {
+
+    val conf = optConf.getOrElse(HBaseConfiguration.create())
+    val connection = optConf.getOrElse(ConnectionFactory.createConnection(conf)).asInstanceOf[Connection]
+    val admin = optAdmin.getOrElse(connection.getAdmin)
+    val hbaseContext = if (optHBaseContext.isDefined) optHBaseContext.get else new HBaseContext(sc, conf)
+
+    /**
+     * Create HBase table
+     *
+     * @param tableDescriptorMeta
+     */
+    private[hbase] def createTable(tableDescriptorMeta: HTableDescriptor) { admin.createTable(tableDescriptorMeta) }
+
+    /**
+     * Get HBase Table
+     *
+     * @param tableName
+     * @return
+     */
+    private[hbase] def getTable(tableName: TableName): Table = { connection.getTable(tableName) }
+
+    /**
+     * Perform HBase query and return data as RDD
+     *
+     * @param tableName
+     * @param scans defines HBase query
+     * @return RDD of HBase Key and Result
+     */
+    private[hbase] def getHBaseRDD(tableName: org.apache.hadoop.hbase.TableName,
+                                   scans: org.apache.hadoop.hbase.client.Scan): RDD[scala.Tuple2[org.apache.hadoop.hbase.io.ImmutableBytesWritable, org.apache.hadoop.hbase.client.Result]] = {
+      hbaseContext.hbaseRDD(tableName, scans)
+
+    }
+
+    /**
+     * Bulk loads data already formatted as Array[Byte] key and values into HBase
+     *
+     * @param genodata
+     * @param hbaseTableName
+     * @param flatMap
+     * @param stagingFolder
+     *
+     * Note:  org.apache.hadoop.hbase.spark.FamilyHFileWriteOptions are currently hard-coded as
+     *        FamilyHFileWriteOptions("GZ", "ROW", 128, "FAST_DIFF")
+     */
+    private[hbase] def hbaseBulkLoad(genodata: RDD[(Array[Byte], List[(String, Array[Byte])])],
+                                     hbaseTableName: String,
+                                     flatMap: ((Array[Byte], List[(String, Array[Byte])])) => Iterator[(Tuple2[KeyFamilyQualifier, Array[scala.Byte]])],
+                                     stagingFolder: String) = {
+      val familyHBaseWriterOptions = new java.util.HashMap[Array[Byte], FamilyHFileWriteOptions]
+      val f1Options = new FamilyHFileWriteOptions("GZ", "ROW", 128, "FAST_DIFF")
+      familyHBaseWriterOptions.put(Bytes.toBytes("g"), f1Options)
+
+      // call implicit RDD function hbaseBulkLoad to write bulk files to staging locations
+      // and then load to HBase
+      genodata.hbaseBulkLoad(hbaseContext,
+        TableName.valueOf(hbaseTableName),
+        flatMap,
+        stagingFolder,
+        familyHBaseWriterOptions,
+        compactionExclude = false,
+        HConstants.DEFAULT_MAX_FILE_SIZE)
+
+      // This permission change appears necessary, plan to revisit to find better way
+
+      val fileSystem = FileSystem.get(conf)
```

Comment:
This is a common (and subtle!) anti-pattern that I see in Hadoop code. Specifically, getting the FileSystem before getting the Path. If you get the Path first, you can then call `path.getFileSystem(conf)`, which works if the path is not on the default filesystem in the hadoop conf files.

Personally, I think it's bad API design in the Hadoop FileSystem class, but that's just my opinion.